x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/common/Failures.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ql.common;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * Glorified list for managing {@link Failure}s.
+ */
+public class Failures {
+
+    private final Collection<Failure> failures;
+
+    public Failures() {
+        this(new ArrayList<>());
+    }
+
+    public Failures(Collection<Failure> failures) {
+        this.failures = failures;
```

Comment:
Slightly dangerous - the `add` method assumes the collection is mutable, but we don't know that the collection passed in here is.  In practice, probably not a huge deal, but we should at least put some java doc on it to indicate the parameter is expected to be mutable.  If we want to be defensive, we could just make a mutable deep copy here too.

Or just don't include this constructor.

Reply:
++, `add` below might have surprisingly different behavior depending on whether we call the constructor with a list or a set of failures.

Maybe `Failures.failures` should be an `ArrayList<Failure>` for simplicity (or a set, if we want set semantics), and we just copy in this constructor?

Reply:
I've encapsulated the code better as the list is not used elsewhere. 
As a side-note, internally the ESQL code should NOT be defensive or paranoid and rely on copying where necessary.